### PR TITLE
get it working on Mac

### PR DESCRIPTION
**Pull request by brlcad**
Hi, kudos on your fun game.  These aren't the cleanest of edits, but thought you might like to see the changes I made to get it working on Mac.  The changes should consequently also still work on Linux, but I didn't test that.

### CODE REVIEW

File: src/main.c [hunk 1]
```diff
@@ -12,10 +12,11 @@
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details. */
 
+#include <unistd.h>
 #include <ncurses.h>
 #include <stdbool.h>
 #include <stdlib.h>
-#include <threads.h>
+#include <pthread.h>
```
Comment (brlcad) [inline src/main.c:19]: I'll note this header isn't needed at all -- I originally thought you were doing threading.

File: src/screen.c [hunk 2]
```diff
@@ -12,11 +12,12 @@
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  * GNU General Public License for more details. */
 
+#include <unistd.h>
 #include <locale.h>
 #include <ncurses.h>
 #include <stdbool.h>
 #include <stdlib.h>
-#include <threads.h>
+#include <pthread.h>
```
Comment (brlcad) [inline src/screen.c:20]: Ditto, not needed.

**Comment by ormai** [comment]
Hey, thanks for the interest!

If we don't link to `ncursesw` it breaks on Debian, like this

image

It seems that MacOS doesn't support `ncursesw` but just `ncurses` which has wide character support enabled by default. Found [this issue](https://github.com/murarth/linefeed/issues/2). Don't know how to solve this.

The other problem seems to relate to the version of the standard used. `<threads.h>` is in c11 with `nanosleep` while `usleep` is deprecated. Can you please try to build with

```sh
make CC=clang CFLAGS=-std=c18
```

Maybe we can solve this by specifying the standard version in the Makefile.

Can you also give me your `clang --version`?

The `[[fallthrough]]`s seemed ok since gcc complained only with `-Wpedantic` and if you give clang `-std=c2x` it works, but we are not there yet.

**Comment by brlcad** [comment]
Yeah, I don't have an easy solution for some of the differences because Make is pretty limited in terms of what it'll do for you. I think most that stick to Make just conditionalize on platform.  If you were using the CMake build system generator, it would handle the difference for you as you'd just run:

set (CURSES_NEED_WIDE 1)
find_package(Curses REQUIRED)

And that would do the trick.  As for the thread.h api, that is not available on Mac.  They use the LLVM Clang compiler (Apple clang version 14.0.3 (clang-1403.0.22.14.1)) and they've thus far said no to that aspect of the C11/C18 changes.  They do define __STDC_NO_THREADS__ accordingly per the standard as the standard doesn't require implementations to provide the threading interface.  It was a contentious issue.

So you could conditionalize the pause method with preprocessor, but usleep is pretty much universal (it's not going away despite deprecation), and has a direct #define corollary on Windows should you ever try to get this working there.  Alternatively, you could use select() as an even older method for inserting a timing pause and that will work everywhere.

That said, adding -std=c11 or -std=c18 to the flags also does not fix the Fallthrough annotation as it is also not provided by the compiler.  It's my understanding that it may be coming for -std=c23 but definitely a no go no mater what -std options are fed.

**Comment by ormai** [comment]
I thought that the standard was... standard. What a mess. I even tried here on Linux with clang and it works, but then Apple... how naive of me.

Given the size of this project I thought that a simple makefile would be appropriate. Setting it up wasn't even that easy, I still don't understand how sub directories are managed.

I know it's not posix, but would an ifeq solve this?
I may also switch to mason altogether.

**Comment by brlcad** [comment]
The standard is an ideal that has never been actually available cross-platform to all platforms, never has been, probably never will.  Not only are there are many to choose from (ISO, GNU, ANSI, IEC, MISRA, POSIX, BSD, ...), and multiple vendors (gcc, clang, msvc, intel, amd, apple, ...), even just getting all the Linux distress to align is impossible.  Latest GCC on Linux is certainly popular, but I could point you at an Enterprise version that still wouldn't be compliant with these newer features.

For what it's worth, Apple is technically compliant.  If memory serves, the standard does not require implementors to adopt the features you're using.  C18 describes them, but also says what to do when you don't want to provide them.  Several implementors did not because pthreads are vastly superior to what C18 ultimately defined.  They're also not alone in that regard.  You'll find similar incompatibility if you look at other non-Linux or enterprise Linux environments.

Mason would be a solution.  CMake would also be a solution.  Both are pretty popular and simple to set up.  CMake's syntax and conventions really suck, but it is by far the most popular for true cross-platform.  The syntax is not really any more complex than what you already have in Make, especially if you added support for one more environment.

**Comment by ormai** [comment]
I did not know it was so complicated for C. This [this xkcd](https://xkcd.com/927/) has been accurate all along.

I think I got it working with meson. I'll merge this, then do the switch.

**Pull request merged**